lib/mfg.c

```diff
@@ -149,39 +149,60 @@ struct get_cfgs_reply {
 	uint8_t rsvd4[32];
 };
 
+struct get_cfgs_reply_gen5 {
+	uint32_t valid0;
+	uint32_t valid1;
+	uint64_t cfg;
+	uint32_t public_key_exponent;
+	uint8_t rsvd2;
+	uint8_t public_key_num;
+	uint8_t public_key_ver;
+	uint8_t spi_core_clk_high;
+	uint8_t public_key[10][SWITCHTEC_KMSK_LEN];
+	uint32_t cdi_efuse_inc_mask;
+	uint8_t uds_data[32];
+};
+
 static int get_configs(struct switchtec_dev *dev,
 		       struct get_cfgs_reply *cfgs,
 		       int *otp_valid)
 {
 	uint8_t subcmd = 0;
 	int ret;
 
-	if (switchtec_gen(dev) == SWITCHTEC_GEN5) {
-		subcmd = 1;
-		ret = switchtec_mfg_cmd(dev,
-					MRPC_SECURITY_CONFIG_GET_GEN5,
-					&subcmd, sizeof(subcmd),
-					cfgs, sizeof(struct get_cfgs_reply));
-		if (!ret)
-			*otp_valid = true;
-	} else {
-		ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_GET_EXT,
-					&subcmd, sizeof(subcmd),
-					cfgs, sizeof(struct get_cfgs_reply));
-		if (ret && ERRNO_MRPC(errno) != ERR_CMD_INVALID)
-			return ret;
-
-		if (!ret) {
-			*otp_valid = true;
-			return ret;
-		}
+	ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_GET_EXT,
+				&subcmd, sizeof(subcmd),
+				cfgs, sizeof(struct get_cfgs_reply));
+	if (ret && ERRNO_MRPC(errno) != ERR_CMD_INVALID)
+		return ret;
 
-		*otp_valid = false;
-		ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_GET,
-					NULL, 0, cfgs,
-					sizeof(struct get_cfgs_reply));
+	if (!ret) {
+		*otp_valid = true;
+		return ret;
 	}
 
+	*otp_valid = false;
+	ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_GET,
+				NULL, 0, cfgs,
+				sizeof(struct get_cfgs_reply));
+
+	return ret;
+}
+
+static int get_configs_gen5(struct switchtec_dev *dev,
+			    struct get_cfgs_reply_gen5 *cfgs,
+			    int *otp_valid)
+{
+	uint32_t subcmd = 0;
+	int ret;
+
+	ret = switchtec_mfg_cmd(dev,
+				MRPC_SECURITY_CONFIG_GET_GEN5,
+				&subcmd, sizeof(subcmd),
+				cfgs, sizeof(struct get_cfgs_reply_gen5));
+	if (!ret)
+		*otp_valid = true;
```

Comment:
It seems opt_valid is always true when we are able to get the cfgs, then we probably don't need the opt_valid as an out parameter.